### ISSUE TITLE
Background Service stops when the user closes the application (iPhone).

### ISSUE THREAD

**Issue by shernandezp** [shernandezp/XamarinForms.LocationService#1]

**Comment by gauravkundalwal** [comment]
Hello,

I tried running this application as it is on my mobile but it stops after 7-8 minutes. I locked my phone after pressing starting the service.
Any help on it would be great.

**Comment by shernandezp** [comment]
Mmm, that sounds something related to Foreground permissions in your device; even I added those permissions to the Manifest and info.plist files, you would better to check them after installing the app. I haven't tested the app with Android 11, the rest of the versions Android and iPhone in general didn't show that issue.

**Comment by gauravkundalwal** [comment]
Here are the files and allowed permission, these are all checked. I also checked this on Android Simulator Pixel2 (API 28) and it gets close. Am I missing something? 

![AllowedPermissions](https://user-images.githubusercontent.com/72265411/94981998-bced6b80-0554-11eb-84ce-763996bf557c.jpg)
![AtInstallPermissionAsk](https://user-images.githubusercontent.com/72265411/94982000-c080f280-0554-11eb-9b3f-5b74d613c442.jpg)
![AllowedPermissionManifest](https://user-images.githubusercontent.com/72265411/94982006-cbd41e00-0554-11eb-8c7e-71a981b766b5.png)

**Comment by shernandezp** [comment]
![Notification1](https://user-images.githubusercontent.com/13307854/94982976-fd28fb80-050c-11eb-83ba-eb5b42e617ab.jpg)
![Notification2](https://user-images.githubusercontent.com/13307854/94982977-fdc19200-050c-11eb-8517-a0cc9af624ac.jpg)

Mmm, you are not getting the notification.

**Comment by gauravkundalwal** [comment]
Thanks, Its was helpful. 

After your message I ran it on Android Simulator and it worked.  Also showing Notification Icon on the status bar but still doesn't work on my phone and doesn't show same notification logo on the status bar.

I have Vivo V11 Pro. Any clue where I should look into to troubleshoot it?

**Comment by shernandezp** [comment]
You could debug the app on your phone and see what's going in the NotificationHelper class; even before that, in the method OnStartCommand of the AndroidLocationService class.

**Comment by gauravkundalwal** [comment]
I kept looking for "Foreground Services" related issues on Chinese Manufacturer devices and Samsung and it turned out that they maintain WhiteList approach for running  Foreground/Background services to enhance battery life on their devices. Company i.e. Facebook, Google already has such app categorized as Whitelist apps for their OS.

Then I thought to give it a try and I deeply looked into my phone for every single setting related to Background Running Process and for Notifications. I also checked Battery Settings. 

And the results was after enabling your app "XamarinForms.Location" for "High background battery usage", it was able to run in background even when the screen was locked.

I am digging into it more and will back.

**Comment by ormesam** [comment]
Running this app on a Samsung A40 worked fine and ran without issues for over half an hour until I stopped it. During that time I also ran it on my Huawei P20 Pro and it lasted less than 10 minutes each time before the OS killed it off. Both devices were unplugged and had the same amount of charge (trying to debug this is a real pain as foreground services will not be killed while the phone is plugged in) and also used the default power saving settings.

Huawei is particularly bad for killing off services. There does seem to be some sort of work around for this though, check out this [StackOverflow Question](https://stackoverflow.com/questions/48166206/how-to-start-power-manager-of-all-android-manufactures-to-enable-background-and).

**Comment by devsuperman** [comment]
Hiiiiiiiiii guuuuuuuuyss!!

Very very very so much wonderful thanks gauravkundalwal !!! 

I have a Xiaomi Redmi Note 8 and the service mysteriously stops when I open other apps. After disabling the Xiaomi Battery Saver, the location service worked perfectly!

1. I went to App Info
2. Enter in Battery Saver
3. The option selected was "Baterry Saver", so i select the "No Restrictions" option

Finish!

**Comment by Hiden46** [comment]
I have ported part of this project to my own application and am testing it. On an Android Honor 10 it seemed to me that if I start Google Maps and start navigation, the service stops. I tried to restart it and it went for a while but within 30 minutes the service stopped again (I don't know exactly at what minute because I was driving)

**Comment by paulCrp** [comment]
Hello @Hiden46 i test this app in many config and i'm going to share my result if it's can help. On Android 10 you have multiple restrictions. It's depend of the phone and constructor. I've made several test and it look like you can't go under 25 second location retrieval (approximately). If the service ask to many times, the security of the OS kill the app responsible for that. In addition of setting a larger timewindow for retrieval, you can also try this solution : https://stackoverflow.com/a/67291554/11630806 . But even with that, the app stop in 2 hour if the phone is not plug (on android 10 only and it's depends of smartphone brand). Also, if you use another app like googleMap or Waze for exemple, the conflict between the multiple call to the location ended you're app. Hope it helps, if you find better solutions or if you have other result in test don't hesitate to share it.

**Comment by Hiden46** [comment]
Hello @paulCrp I have left the time to request the GPS signal exactly as in the app above, every 2 seconds I make a request to get the coordinates, speed and location of the device.

**Comment by ormesam** [comment]
Correct me if I'm wrong but it looks like android:foregroundServiceType="location" is missing from the service / AndroidManifest? Some Android devices will kill the foreground service after a while if it is not told what it does. https://developer.android.com/guide/topics/manifest/service-element

**Comment by Hiden46** [comment]
Hi @ormesam I tried as you wrote to put the <service> tag in the manifest, referring to the service inside the Android part, but nothing has changed. I am currently doing some tests and shortly after the application goes in the background it stops getting the gps signal. It starts over as I reopen it.

**Comment by ormesam** [comment]
Hi @Hiden46 did you add the [Register("...")] attribute to the service too?

This is how I do it in one of my projects:

**Service**
` [Register("com.samorme.tracked.ForegroundService")]    public class ForegroundService : Service {...}`

**AndroidManifest**
```

    <?xml version="1.0" encoding="utf-8"?>
    <manifest xmlns:android="http://schemas.android.com/apk/res/android" android:versionCode="14" android:versionName="1.4" package="com.samorme.tracked" android:installLocation="internalOnly">
        <uses-sdk android:minSdkVersion="29" android:targetSdkVersion="29">
            <application>
                <service android:enabled="true" android:foregroundServiceType="location" android:exported="false" android:name="com.samorme.tracked.ForegroundService" />
            </application>
    </manifest>
```

**Comment by Hiden46** [comment]
@ormesam Yes I did. I have noticed that if I try using visual studio debugging with cable phone connected the service continues to work. When I try the app instead without being debugged with visual studio and unplugged the phone from the PC it stops working after a while. The active service icon remains, but you notice that the GPS system is not working because the icon does not appear in the status bar.

I put the _Register_ in the `AndroidLocationService.cs` un Android project in the Services folder
`[Register("com.dolomicchiostreetapp.TrackingService")] public class AndroidLocationService : Service{ ... }`
